lib/remote-bitfield.js

```diff
@@ -1,63 +1,224 @@
 const BigSparseArray = require('big-sparse-array')
 const quickbit = require('quickbit-universal')
 
+const BITS_PER_PAGE = 32768
+const BYTES_PER_PAGE = BITS_PER_PAGE / 8
+const BITS_PER_SEGMENT = 2097152
+const BYTES_PER_SEGMENT = BITS_PER_SEGMENT / 8
+const PAGES_PER_SEGMENT = BITS_PER_SEGMENT / BITS_PER_PAGE
+
+class RemoteBitfieldPage {
+  constructor (index, bitfield, segment) {
+    this.index = index
+    this.bitfield = bitfield
+    this.segment = segment
+    this.offset = index * BYTES_PER_PAGE - segment.index * BYTES_PER_SEGMENT
+
+    segment.add(this)
+  }
+
+  get tree () {
+    return this.segment.tree
+  }
+
+  get (index) {
+    return quickbit.get(this.bitfield, index)
+  }
+
+  set (index, val) {
+    if (quickbit.set(this.bitfield, index, val)) {
+      this.tree.update(this.offset * 8 + index)
+    }
+  }
+
+  setRange (start, length, val) {
+    quickbit.fill(this.bitfield, val, start, start + length)
+
+    let i = Math.floor(start / 32)
+    const n = i + Math.ceil(length / 32)
+
+    while (i < n) this.tree.update(this.offset * 8 + i++ * 32)
+  }
+
+  findFirst (val, position) {
+    return quickbit.findFirst(this.bitfield, val, this.tree.skipFirst(!val, this.offset * 8 + position))
+  }
+
+  findLast (val, position) {
+    return quickbit.findLast(this.bitfield, val, this.tree.skipLast(!val, this.offset * 8 + position))
+  }
+
+  insert (start, bitfield) {
+    this.bitfield.set(bitfield, start / 32)
+  }
+}
+
+class RemoteBitfieldSegment {
+  constructor (index) {
+    this.index = index
+    this.tree = quickbit.Index.from([])
+  }
+
+  get chunks () {
+    return this.tree.chunks
+  }
+
+  add (page) {
+    let j = -1
+    let i = this.chunks.length
+
+    while (j + 1 < i) {
+      const m = j + ((i - j) >>> 1)
+
+      if (page.index < this.chunks[m].index) {
+        i = m
+      } else {
+        j = m
+      }
+    }
+
+    this.chunks.splice(i, 0, { field: page.bitfield, offset: page.offset })
```

Comment:
since the splice is linear anyway you can just do this like so

```js
chunks.push(x)
for (let i = chunks.length - 2; i >= 0; i--) {
  const prev = chunks[i]
  if (prev <= x) break
  chunks[i] = x
  chunks[i + 1] = prev
} 
```